### DISCUSSION TITLE
uAgents need ability to store environment variables

### DISCUSSION THREAD

**Discussion by jasonsprouse** [fetchai/uAgents#88] [General]
We need to be able to use a passphrase to spawn an account, but we're also going to need to be able to input api keys for other products.

For example, with something like [JARVIS](https://github.com/microsoft/JARVIS#quick-start), a LLM being built that utilizes more refined models that process through a task que, we will need environment variables to store the api keys.

**Comment by jasonsprouse** [comment]
Primarily in a panel in the Agentverse where we would add the repo from github and input the variables needed for the agent.

**Comment by jrriehl** [comment]
Thanks for raising this request. We are already considering how to store environment variables, and we will soon publish milestones for greater transparency on the development of new features.

**Comment by jrriehl** [comment]
@lrahmani please update this with the milestone once it is ready.

**Comment by abhifetch** [comment]
@lrahmani do you have any update on this?

**Comment by abhifetch** [comment]
You can store API keys in agent's secret and it will be stored securely as a variable. for more details refer: 
<img width="985" alt="Screenshot 2024-08-20 at 09 43 32" src="https://github.com/user-attachments/assets/0193c171-2492-4a86-8906-36975bd5fadd">
You can add any number of secrets here and use that as variable in your agent's script.

**Comment by musaabhasan** [comment]
Environment variables are useful for configuration, but secrets should be treated as a separate class of data.

A practical model would be:

- normal env vars for non-sensitive runtime configuration
- secret references for API keys, passphrases, tokens, and private endpoints
- encrypted storage at rest
- runtime injection only into the specific agent execution context that needs it
- masking in UI, logs, traces, and exports
- per-secret ownership, rotation, and audit history

For agents, I would also avoid making every secret globally available to the whole agent. If an agent has multiple tools, the secret should be scoped to the tool or capability that needs it. Otherwise a prompt-injection or tool-confusion bug can turn one integration key into authority for unrelated actions.

A good UX detail would be a manifest-driven setup panel: the agent declares required secret names and descriptions, and Agentverse prompts the user to supply values without the repo ever containing them.

**Comment by jasonsprouse** [comment]
The reason this was posted, was at the time this was addressed there was no place for secrets to be stored privately. All data would be exposed. Environment variables are secrets --- if you don't need secrets, you can just hard code whatever your app requires into the app.

Best Regards,



Jason Sprouse

306 Franklin

Pottsboro, TX 75076

Executive Director

Good Fatih Paradigm<https://goodfaith.church/>
p. 469-844-0275
c. 903-357-9559

Please consider the environment before printing this email. This email and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are addressed. If you have received this email in error please notify the sender immediately. Please note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of Good Faith Paradigm. Finally, the recipient should check this email and any attachments for the presence of viruses. The company accepts no liability for any damage caused by any virus transmitted by this email.

________________________________
